File: Validation/Phase1_VTilted_Catchment/generate_phase1_vtilted_domain.py
#!/usr/bin/env python3
"""Generate the Phase 1 v-tilted synthetic catchment rasters.

This is the Python equivalent of `generate_phase1_vtilted_domain.m`, provided
so the domain can be generated in environments without MATLAB.
"""

from __future__ import annotations

import csv
from pathlib import Path

import numpy as np
import rasterio
from rasterio.transform import from_origin


def read_config(path: Path) -> dict[str, str]:
    with path.open(newline="", encoding="utf-8") as handle:
        return {row["parameter"]: row["value"] for row in csv.DictReader(handle)}


def f(config: dict[str, str], name: str) -> float:
    return float(config[name])


def write_raster(path: Path, data: np.ndarray, transform, epsg: int) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with rasterio.open(
        path,
        "w",
        driver="GTiff",
        height=data.shape[0],
        width=data.shape[1],
        count=1,
        dtype="float32",
        crs=f"EPSG:{epsg}",
        transform=transform,
        nodata=None,
        compress="lzw",
    ) as dst:
        dst.write(data.astype("float32"), 1)


def write_csv(path: Path, rows: list[dict[str, object]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(rows[0].keys()))
        writer.writeheader()
        writer.writerows(rows)


def main() -> int:
    case_dir = Path(__file__).resolve().parent
    static_dir = case_dir / "Static"
    output_dir = case_dir / "Outputs" / "Validation"
    config = read_config(case_dir / "Config" / "Domain_Config.csv")

    dx = f(config, "dx")
    length_y = f(config, "length_y")
    left_width = f(config, "left_width")
    channel_width = f(config, "channel_width")
    right_width = f(config, "right_width")
    left_slope = f(config, "left_slope")
    right_slope = f(config, "right_slope")
    long_slope = f(config, "longitudinal_slope")
    z_outlet = f(config, "z_outlet")
    channel_incision = f(config, "channel_incision")
    epsg = int(f(config, "epsg"))

    width_total = left_width + channel_width + right_width
    nx = int(round(width_total / dx))
    ny = int(round(length_y / dx))

    col = np.arange(nx, dtype=float)
    row = np.arange(ny, dtype=float)
    x = col[None, :] * dx
    y = row[:, None] * dx

    x_left_end = left_width
    x_channel_end = left_width + channel_width

    left_mask = np.broadcast_to(x < x_left_end, (ny, nx))
    channel_mask = np.broadcast_to((x >= x_left_end) & (x <= x_channel_end), (ny, nx))
    right_mask = np.broadcast_to(x > x_channel_end, (ny, nx))

    dem = z_outlet + long_slope * np.broadcast_to(y, (ny, nx))
    channel_base = dem - channel_incision
    dem[left_mask] = channel_base[left_mask] + left_slope * (x_left_end - np.broadcast_to(x, (ny, nx))[left_mask])
    dem[channel_mask] = channel_base[channel_mask]
    dem[right_mask] = channel_base[right_mask] + right_slope * (np.broadcast_to(x, (ny, nx))[right_mask] - x_channel_end)

    lulc = np.zeros((ny, nx), dtype=float)
    soil = np.zeros((ny, nx), dtype=float)
    zone_id = np.zeros((ny, nx), dtype=float)

    for array in (lulc, soil, zone_id):
        array[left_mask] = 1
        array[channel_mask] = 2
        array[right_mask] = 3

    def zoned(left_name: str, channel_name: str, right_name: str) -> np.ndarray:
        arr = np.zeros((ny, nx), dtype=float)
        arr[left_mask] = f(config, left_name)
        arr[channel_mask] = f(config, channel_name)
        arr[right_mask] = f(config, right_name)
        return arr

    dtb = zoned("dtb_left", "dtb_channel", "dtb_right")
    albedo = zoned("albedo_left", "albedo_channel", "albedo_right")
    lai = zoned("lai_left", "lai_channel", "lai_right")
    initial_sm = zoned("initial_sm_left", "initial_sm_channel", "initial_sm_right")
    gw_depth = zoned("gw_depth_left", "gw_depth_channel", "gw_depth_right")
    gw_table = dem - dtb + gw_depth

    rasters = {
        "DEM.tif": dem,
        "LULC.tif": lulc,
        "SOIL.tif": soil,
        "DTB.tif": dtb,
        "Albedo.tif": albedo,
        "LAI.tif": lai,
        "Initial_SM.tif": initial_sm,
        "GW_depth.tif": gw_depth,
        "GW_table.tif": gw_table,
        "Zone_ID.tif": zone_id,
    }

    for name, arr in rasters.items():
        if not np.isfinite(arr).all():
            raise ValueError(f"{name} contains NaN or Inf")

    transform = from_origin(0.0, length_y, dx, dx)
    for name, arr in rasters.items():
        write_raster(static_dir / name, arr, transform, epsg)

    summary_rows = [
        {
            "zone": "left_hillslope",
            "cell_count": int(left_mask.sum()),
            "lai": f(config, "lai_left"),
            "initial_sm_mm": f(config, "initial_sm_left"),
            "dtb_m": f(config, "dtb_left"),
        },
        {
            "zone": "channel_strip",
            "cell_count": int(channel_mask.sum()),
            "lai": f(config, "lai_channel"),
            "initial_sm_mm": f(config, "initial_sm_channel"),
            "dtb_m": f(config, "dtb_channel"),
        },
        {
            "zone": "right_hillslope",
            "cell_count": int(right_mask.sum()),
            "lai": f(config, "lai_right"),
            "initial_sm_mm": f(config, "initial_sm_right"),
            "dtb_m": f(config, "dtb_right"),
        },
    ]
    write_csv(output_dir / "VTilted_Domain_Summary.csv", summary_rows)

    print(f"Generated {len(rasters)} rasters in {static_dir}")
    print(f"Wrote summary to {output_dir / 'VTilted_Domain_Summary.csv'}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Validation/Phase1_VTilted_Catchment/generate_phase1_vtilted_domain.m
%% Phase 1 V-tilted synthetic catchment generator
% Generates static rasters for the shared Phase 1 validation domain.
%
% Run from the repository root:
%   run('HydroPol2D_Model/Validation/Phase1_VTilted_Catchment/generate_phase1_vtilted_domain.m')

clear; clc;

case_dir = fileparts(mfilename('fullpath'));
static_dir = fullfile(case_dir, 'Static');
config_file = fullfile(case_dir, 'Config', 'Domain_Config.csv');

if ~exist(static_dir, 'dir')
    mkdir(static_dir);
end

cfg = readtable(config_file, 'TextType', 'string');
getv = @(name) cfg.value(cfg.parameter == name);
getn = @(name) str2double(getv(name));

dx = getn('dx');
lengthY = getn('length_y');
leftWidth = getn('left_width');
channelWidth = getn('channel_width');
rightWidth = getn('right_width');
leftSlope = getn('left_slope');
rightSlope = getn('right_slope');
longSlope = getn('longitudinal_slope');
z_outlet = getn('z_outlet');
channelIncision = getn('channel_incision');

DTB_left = getn('dtb_left');
DTB_channel = getn('dtb_channel');
DTB_right = getn('dtb_right');

Albedo_left = getn('albedo_left');
Albedo_channel = getn('albedo_channel');
Albedo_right = getn('albedo_right');

LAI_left = getn('lai_left');
LAI_channel = getn('lai_channel');
LAI_right = getn('lai_right');

InitialSM_left = getn('initial_sm_left');
InitialSM_channel = getn('initial_sm_channel');
InitialSM_right = getn('initial_sm_right');

GWdepth_left = getn('gw_depth_left');
GWdepth_channel = getn('gw_depth_channel');
GWdepth_right = getn('gw_depth_right');
epsgCode = getn('epsg');

widthTotal = leftWidth + channelWidth + rightWidth;
nx = round(widthTotal / dx);
ny = round(lengthY / dx);

[X, Y] = meshgrid(1:nx, 1:ny);
x = (X - 1) * dx;
y = (Y - 1) * dx;

activeMask = true(ny, nx);

x_left_end = leftWidth;
x_channel_end = leftWidth + channelWidth;

leftMask = activeMask & x < x_left_end;
channelMask = activeMask & x >= x_left_end & x <= x_channel_end;
rightMask = activeMask & x > x_channel_end;

DEM = z_outlet + longSlope .* y;
channelBase = DEM - channelIncision;

DEM(leftMask) = channelBase(leftMask) + leftSlope .* (x_left_end - x(leftMask));
DEM(channelMask) = channelBase(channelMask);
DEM(rightMask) = channelBase(rightMask) + rightSlope .* (x(rightMask) - x_channel_end);

LULC = zeros(ny, nx);
SOIL = zeros(ny, nx);
Zone_ID = zeros(ny, nx);

LULC(leftMask) = 1;
LULC(channelMask) = 2;
LULC(rightMask) = 3;

SOIL(leftMask) = 1;
SOIL(channelMask) = 2;
SOIL(rightMask) = 3;

Zone_ID(leftMask) = 1;
Zone_ID(channelMask) = 2;
Zone_ID(rightMask) = 3;

DTB = zeros(ny, nx);
Albedo = zeros(ny, nx);
LAI = zeros(ny, nx);
Initial_SM = zeros(ny, nx);
GW_depth = zeros(ny, nx);

DTB(leftMask) = DTB_left;
DTB(channelMask) = DTB_channel;
DTB(rightMask) = DTB_right;

Albedo(leftMask) = Albedo_left;
Albedo(channelMask) = Albedo_channel;
Albedo(rightMask) = Albedo_right;

LAI(leftMask) = LAI_left;
LAI(channelMask) = LAI_channel;
LAI(rightMask) = LAI_right;

Initial_SM(leftMask) = InitialSM_left;
Initial_SM(channelMask) = InitialSM_channel;
Initial_SM(rightMask) = InitialSM_right;

GW_depth(leftMask) = GWdepth_left;
GW_depth(channelMask) = GWdepth_channel;
GW_depth(rightMask) = GWdepth_right;

GW_table = DEM - DTB + GW_depth;

rasterNames = {'DEM','LULC','SOIL','DTB','Albedo','LAI','Initial_SM','GW_depth','GW_table','Zone_ID'};
for k = 1:numel(rasterNames)
    A = eval(rasterNames{k});
    if any(isnan(A(:)))
        error('%s contains NaN values.', rasterNames{k});
    end
    if any(isinf(A(:)))
        error('%s contains Inf values.', rasterNames{k});
    end
end

x0 = 0;
y0 = lengthY;
R = maprefcells([x0, x0 + nx * dx], [y0 - ny * dx, y0], [ny, nx], ...
    'ColumnsStartFrom', 'north');

geotiffwrite(fullfile(static_dir, 'DEM.tif'), single(DEM), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'LULC.tif'), single(LULC), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'SOIL.tif'), single(SOIL), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'DTB.tif'), single(DTB), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'Albedo.tif'), single(Albedo), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'LAI.tif'), single(LAI), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'Initial_SM.tif'), single(Initial_SM), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'GW_depth.tif'), single(GW_depth), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'GW_table.tif'), single(GW_table), R, 'CoordRefSysCode', epsgCode);
geotiffwrite(fullfile(static_dir, 'Zone_ID.tif'), single(Zone_ID), R, 'CoordRefSysCode', epsgCode);

summary = table( ...
    ["left_hillslope"; "channel_strip"; "right_hillslope"], ...
    [sum(leftMask(:)); sum(channelMask(:)); sum(rightMask(:))], ...
    [LAI_left; LAI_channel; LAI_right], ...
    [InitialSM_left; InitialSM_channel; InitialSM_right], ...
    [DTB_left; DTB_channel; DTB_right], ...
    'VariableNames', {'zone','cell_count','lai','initial_sm_mm','dtb_m'});
writetable(summary, fullfile(case_dir, 'Outputs', 'Validation', 'VTilted_Domain_Summary.csv'));

figure('Color', 'w', 'Position', [100 100 1600 850]);
subplot(2,4,1); imagesc(DEM); axis image off; colorbar; title('DEM [m]');
subplot(2,4,2); imagesc(Zone_ID); axis image off; colorbar; title('Zone ID');
subplot(2,4,3); imagesc(LULC); axis image off; colorbar; title('LULC');
subplot(2,4,4); imagesc(SOIL); axis image off; colorbar; title('SOIL');
subplot(2,4,5); imagesc(DTB); axis image off; colorbar; title('DTB [m]');
subplot(2,4,6); imagesc(Albedo); axis image off; colorbar; title('Albedo');
subplot(2,4,7); imagesc(LAI); axis image off; colorbar; title('LAI');
subplot(2,4,8); imagesc(GW_table); axis image off; colorbar; title('GW table [m]');
sgtitle('Phase 1 V-tilted Catchment Inputs');
exportgraphics(gcf, fullfile(static_dir, 'overview.png'), 'Resolution', 300);

fprintf('Phase 1 v-tilted catchment written to:\n%s\n', case_dir);
